File: be/src/controllers/payment.controller.js
import paymentService from '../services/payment.service.js';
import redisClient from '../dbs/redisdb.js';

export default class paymentController {

    static async checkTransaction(req, res) {
        try {
            const userId = req.user._id;
            const { transactionId, amount } = req.body;
            const hex_key = await redisClient.get(`transaction:${transactionId}`);
            const payload = { userId, transactionId, hex_key, amount };
            const result = await paymentService.checkTransaction(payload);
            res.status(200).json({
                message: 'Check transaction successfully',
                status: result
            });
        } catch (error) {
            res.status(500).json({
                message: error.message || 'Check transaction failed'
            });
        }
    }

    static async createTransaction(req, res) {
        try {
            const userId = req.user._id;
            const transaction = req.body;
            const payload = { userId, ...transaction };
            const newTransaction = await paymentService.createTransaction(payload);
            if (!newTransaction) {
                throw new Error('Create transaction failed');
            }
            const redis_key = `transaction:${newTransaction._id}`;
            await redisClient.set(redis_key, newTransaction.hex_key);
            res.status(200).json({
                message: 'Create transaction successfully',
                transaction: newTransaction
            });
        } catch (error) {
            res.status(500).json({
                message: error.message || 'Create transaction failed'
            });
        }
    }

    static async deleteTransaction(req, res) {
        try {
            const userId = req.user._id;
            const transactionId= req.params.id;
            await paymentService.deleteTransaction({
                userId,
                transactionId
            });
            res.status(200).json({
                message: 'Delete transaction successfully'
            });
        } catch (error) {
            res.status(500).json({
                message: error.message || 'Delete transaction failed'
            });
        }
    }

    static async getTransactionById(req, res) {
        try {
            const userId = req.user._id;
            const { id } = req.params;
            const transactions = await paymentService.getTransactionById({userId, id});
            res.status(200).json({
                message: 'Get transactions successfully',
                transactions
            });
        } catch (error) {
            res.status(500).json({
                message: error.message || 'Get transactions failed'
            });
        }
    }
}

File: BE_JAVA/src/main/java/com/lighting/controller/PaymentController.java
package com.lighting.controller;

import com.lighting.entity.Payment;
import com.lighting.dto.PaymentDto;
import com.lighting.service.PaymentService;
import com.lighting.util.RedisService;
import com.lighting.util.AuthUtil;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import jakarta.validation.Valid;
import java.util.Map;

@RestController
@RequestMapping("/payment")
@RequiredArgsConstructor
@Validated
@CrossOrigin(originPatterns = "*", allowCredentials = "true")
public class PaymentController {

    private static final Logger log = LoggerFactory.getLogger(PaymentController.class);
    private final PaymentService paymentService;
    private final RedisService redisService;

    @PostMapping("/ispaid")
    public ResponseEntity<Map<String, Object>> checkTransaction(@Valid @RequestBody PaymentDto.CheckTransactionRequest request) {
        try {
            String userId = AuthUtil.getCurrentUserId(); // Get from JWT token
            if (userId == null) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", "Unauthorized"));
            }

            String hexKey = redisService.get("transaction:" + request.getTransactionId());
            Map<String, String> result = paymentService.checkTransaction(userId, request.getTransactionId(), hexKey, request.getAmount());

            return ResponseEntity.ok(Map.of(
                "message", "Check transaction successfully",
                "status", result.get("status")
            ));
        } catch (Exception e) {
            log.error("Check transaction failed", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(Map.of("message", e.getMessage() != null ? e.getMessage() : "Check transaction failed"));
        }
    }

    @PostMapping("/create")
    public ResponseEntity<Map<String, Object>> createTransaction(@Valid @RequestBody PaymentDto.CreateTransactionRequest request) {
        try {
            String userId = AuthUtil.getCurrentUserId(); // Get from JWT token
            if (userId == null) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", "Unauthorized"));
            }

            Payment newTransaction = paymentService.createTransaction(userId, request.getAmount());

            if (newTransaction == null) {
                throw new RuntimeException("Create transaction failed");
            }

            // Set Redis key like in Node.js
            String redisKey = "transaction:" + newTransaction.getId();
            redisService.set(redisKey, newTransaction.getHexKey());

            return ResponseEntity.ok(Map.of(
                "message", "Create transaction successfully",
                "transaction", newTransaction
            ));
        } catch (Exception e) {
            log.error("Create transaction failed", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(Map.of("message", e.getMessage() != null ? e.getMessage() : "Create transaction failed"));
        }
    }

    @PostMapping("/delete/{id}")
    public ResponseEntity<Map<String, Object>> deleteTransaction(@PathVariable String id) {
        try {
            String userId = AuthUtil.getCurrentUserId(); // Get from JWT token
            if (userId == null) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", "Unauthorized"));
            }

            paymentService.deleteTransaction(userId, id);

            return ResponseEntity.ok(Map.of(
                "message", "Delete transaction successfully"
            ));
        } catch (Exception e) {
            log.error("Delete transaction failed", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(Map.of("message", e.getMessage() != null ? e.getMessage() : "Delete transaction failed"));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> getTransactionById(@PathVariable String id) {
        try {
            String userId = AuthUtil.getCurrentUserId(); // Get from JWT token
            if (userId == null) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", "Unauthorized"));
            }

            Payment transactions = paymentService.getTransactionById(userId, id);

            return ResponseEntity.ok(Map.of(
                "message", "Get transactions successfully",
                "transactions", transactions
            ));
        } catch (Exception e) {
            log.error("Get transactions failed", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(Map.of("message", e.getMessage() != null ? e.getMessage() : "Get transactions failed"));
        }
    }
}
